community/kernel/src/main/java/org/neo4j/kernel/impl/locking/community/RWLock.java

```diff
@@ -412,12 +422,19 @@ synchronized boolean acquireWriteLock( Object tx ) throws DeadlockDetectedExcept
         }
     }
 
-    private boolean waitUninterruptedly()
+    private boolean waitUninterruptedly( long lockAcquisitionTimeBoundary )
     {
         boolean addLockRequest;
         try
         {
-            wait();
+            if ( lockAcquisitionTimeoutMillis > 0 )
+            {
+                wait( Math.abs( lockAcquisitionTimeBoundary - clock.millis() ) );
```

Comment:
If `clock.millis() >>> lockAcquisitionTimeBoundary` due to thread scheduling or smth, thread will still wait for a while and then throw exception. Would it make sense to make previous condition
`if (lockAcquisitionTimeoutMillis > 0 && lockAcquisitionTimeBoundary > currentMillis)`
and then just `wait(lockAcquisitionTimeBoundary-currentMillis)` ?